crates/core/tedge_actors/src/message_boxes.rs

```diff
@@ -0,0 +1,265 @@
+use crate::Builder;
+use crate::ChannelError;
+use crate::DynSender;
+use crate::Message;
+use crate::MessageBoxConnector;
+use crate::SimpleMessageBoxBuilder;
+use async_trait::async_trait;
+use futures::channel::mpsc;
+use futures::StreamExt;
+use log::debug;
+use log::info;
+
+/// A message box used by an actor to collect all its input and forward its output
+///
+/// This message box can be seen as two streams of messages,
+/// - inputs sent to the actor and stored in the message box awaiting to be processed,
+/// - outputs sent by the actor and forwarded to the message boxes of other actors.
+///
+/// ```logical-view
+///                                      +------+
+/// input_sender: DynSender<Input> ----->| Box  |----> output_sender: DynSender<Input>
+///                                      +------+
+/// ```
+///
+/// Under the hood, a `MessageBox` implementation can use
+/// - several input channels to await messages from specific peers
+///   .e.g. awaiting a response from an HTTP actor
+///    and ignoring the kind of events till a response or a timeout has been received.
+/// - several output channels to send messages to specific peers.
+/// - provide helper function that combine internal channels.
+#[async_trait]
+pub trait MessageBox: 'static + Sized + Send + Sync {
```

Comment:
Any preference for a name: `MessageBox` or `MailBox`?

Reply:
Isn't it a queue? Because they are processed on the incoming order?

Reply:
> Isn't it a queue? Because they are processed on the incoming order?

There are queues under the hood, yes. But I would not stress that in the name, because a message box can use several queues to control prioritization of message ingestion (say to process a runtime requests before any regular request or the await for a response to a pending request ignoring other events till this response is received).

Reply:
We keep the MessageBox name